File: sorting/quick_sort/quick_sort.ts
interface Initial {
  more: number[];
  less: number[];
}

const quick_sort = (list: number[]): number[] => {
  if (list.length < 2) return list;

  const pivot: number = list[0];
  const initial: Initial = { more: [], less: [] };

  const { more, less } = list.slice(1).reduce((acc, n) => {
    n > pivot ? acc.more.push(n) : acc.less.push(n);
    return acc;
  }, initial);

  return quick_sort(less).concat([pivot], quick_sort(more));
};

const list = Array.from({ length: 10 }, () => Math.floor(Math.random() * 40));
const sorted = quick_sort(list);

console.log(sorted);

File: sorting/quick_sort/quick_sort.rb
def quick_sort(list)
  return list if list.size < 2

  pivot = list.first
  more, less = list[1..-1].partition(&->(n) { n > pivot })

  quick_sort(less).concat [pivot], quick_sort(more)
end

pp quick_sort [*1..10].shuffle
